ouroboros-consensus/test/consensus-test/Test/Consensus/Peras/Voting/Rules.hs

```diff
@@ -0,0 +1,313 @@
+-- on a simplified model that doesn't use anything fancy to evaluate the rules.
+prop_isPerasVotingAllowed :: Property
+prop_isPerasVotingAllowed = forAll genPerasVotingView $ \pvv -> do
+  -- Determine whether we should vote according to the model
+  let (shouldVote, votingRules@(vr1a, vr1b, vr2a, vr2b)) =
+        isPerasVotingAllowedModel pvv
+  -- Some helper functions to report success/failure
+  let chain = flip (foldr ($)) . reverse
+  let ok desc =
+        chain
+          [ tabulate "VR-1A" [show vr1a]
+          , tabulate "VR-1B" [show vr1b]
+          , tabulate "VR-2A" [show vr2a]
+          , tabulate "VR-2B" [show vr2b]
+          , tabulate "VR-(1A|1B|2A|2B)" [show votingRules]
+          , tabulate "Should vote according to model" [show shouldVote]
+          , tabulate "Actual result" [desc]
+          ]
+          $ property True
+  let failure desc =
+        counterexample desc $
+          property False
+  -- Now check that the real implementation agrees with the model
+  case isPerasVotingAllowed pvv of
+    Right voteReason
+      | shouldVote ->
+          ok $ "VoteReason(" <> explainShallow voteReason <> ")"
+      | otherwise ->
+          failure $ "Expected not to vote, but got: " <> show voteReason
+    Left noVoteReason
+      | not shouldVote ->
+          ok $ "NoVoteReason(" <> explainShallow noVoteReason <> ")"
+      | otherwise ->
+          failure $ "Expected to vote, but got: " <> show noVoteReason
+
+{-------------------------------------------------------------------------------
+  Arbitrary helpers
+-------------------------------------------------------------------------------}
+
+-- * Peras parameters
+
+-- NOTE: we use a geometric distribution to bias towards smaller values.
+-- This increases the chance of covering all the voting rules more evenly,
+-- while still allowing for larger values to be generated occasionally.
+--
+-- Moreover, geometric(0.5) + 1 means that:
+--  - 50% chance of being 1
+--  - 25% chance of being 2
+--  - 12.5% chance of being 3
+--  ... and so on
+genPerasParams :: Gen PerasParams
+genPerasParams = do
+  _L <- fromIntegral . (+ 1) <$> geometric 0.5
+  _X <- fromIntegral . (+ 1) <$> geometric 0.5
+  _R <- fromIntegral . (+ 1) <$> geometric 0.5
+  _K <- fromIntegral . (+ 1) <$> geometric 0.5
+  pure
+    mkPerasParams
+      { perasBlockMinSlots = PerasBlockMinSlots _L
+      , perasCertArrivalThreshold = PerasCertArrivalThreshold _X
+      , perasIgnoranceRounds = PerasIgnoranceRounds _R
+      , perasCooldownRounds = PerasCooldownRounds _K
+      }
+
+-- * Slot numbers
+
+genSlotNo :: Gen SlotNo
+genSlotNo = do
+  n <- arbitrary
+  pure (SlotNo n)
+
+-- * Peras round numbers
+
+genPerasRoundNo :: Gen PerasRoundNo
+genPerasRoundNo = do
+  n <- arbitrary
+  pure (PerasRoundNo n)
+
+-- * Mocked certificate type
+
+-- NOTE: we could also use the real 'WithArrivalTime (ValidatedPerasCert blk)'
+-- here. However, this one is much easier to derive a 'Function' instance for,
```

Comment:
Does the justification still holds now that the view is pure? Otherwise we should probably use the official type :)